ppa/archive/management/commands/gensim_serialize.py

```diff
@@ -0,0 +1,292 @@
+
+        # list of strings, populated on first doc retrieval
+        self.metadata_field_names = None
+
+        results = self.client.query(
+            collection=self.collection,
+            query={
+                'q': '*:*',
+                'facet': 'on',
+                'rows': 0,
+                'facet.field': doc_id_field,
+                'facet.limit': self.doc_limit
+            }
+        )
+
+        """
+        An OrderedDict of doc_id => page count mapping
+        An OrderedDict is important here in case we want to save document-level metadata, in which case rows of
+        metadata would be in the same order as the BoW-vectors returned by this object's iterator.
+        """
+        self.page_counts = results.get_facets()['source_id']
+        self.doc_ids = doc_ids or self.page_counts.keys()
+        self.doc_count = len(self.doc_ids)
+        self.pbar = ProgressBar(redirect_stderr=True, max_value=self.doc_count) if pbar else NullBar()
+
+    def __iter__(self):
+
+        for doc_id in self.doc_ids:
+
+            if doc_id not in self.page_counts:
+                logger.warning('Unknown page count for doc {}. Skipping.'.format(doc_id))
+                continue
+
+            result = self.client.query(
+                collection=self.collection,
+                query={
+                    'q': '{}:{}'.format(self.doc_id_field, doc_id),
+                    'order': '{} asc'.format(self.page_order_field),
+                    'rows': self.page_counts[doc_id]
+                }
+            )
+
+            metadata_docs = [d for d in result.docs if d['item_type'] == 'work']
+            n_metadata_docs = len(metadata_docs)
+            if n_metadata_docs > 0:
+                if n_metadata_docs > 1:
+                    logger.warning('Multiple metadata records found for doc ID {}. Using the first.'.format(doc_id))
+
+                metadata_doc = metadata_docs[0]
+                self._metadata[doc_id] = {k: v for k, v in metadata_doc.items()}
+                if self.metadata_field_names is None:
+                    self.metadata_field_names = list(metadata_doc.keys())
+            else:
+                logger.warning('No metadata record found for doc ID {}.'.format(doc_id))
+
+            # filter out ages that have no content; combine all pages into one string
+            fulltext = ' '.join(list(
+                filter(
+                    None,
+                    (doc.get(self.doc_content_field) for doc in result.docs if doc['item_type'] == 'page')
+                )
+            ))
+            fulltext = preprocess_string(fulltext, filters=self.preprocess_fns)
+
+            yield self.dictionary.doc2bow(fulltext, allow_update=True)
+            self.pbar.update(self.pbar.value + 1)
+
+    def _save_dictionary(self, filepath):
+        with open(filepath, 'w', encoding='utf8') as f:
+            for i in range(len(self.dictionary)):
+                f.write(self.dictionary[i] + '\n')
```

Comment:
I was trying to remember why I did this..

I had been trying to load the model (.mm) and the dictionary (.txt) in the context of the R `stm` package to get some interesting results. It all works fine, but while the document-term matrix has a nice standardized form (the mm format) that can operate between Python <-> R, there is no agreed-upon format (none that I'm aware of) for the dictionary. Gensim simply pickles the Python object as-is. So I came up with a bare-minimum format for the dictionary (where the 0-indexed line numbers corresponds to the token id). This can be fed into packages like `stm` fairly effortlessly.

But I agree that the dictionary has to be useful in the `gensim` world too. Can we allow for saving in both formats? (with the gensim-aware format being the default). Or perhaps someone with more experience in NLP can suggest a standard format?

Reply:
Interesting! Yes, it's important to have the dictionary be useful with other tools.  Maybe generate both versions for now and we can do research and/or consult someone with NLP expertise to find a more standard format to use in future versions?